test/test_filepost.py

```diff
@@ -1,15 +1,49 @@
-import unittest
-
-from urllib3.filepost import encode_multipart_formdata, iter_fields
+import psutil
+import io
+import sys
+
+if sys.version_info >= (2, 7):
+    import unittest
+else:
+    import unittest2 as unittest
+
+from urllib3.filepost import (
+    encode_multipart_formdata,
+    iter_fields,
+    MultipartEncoderGenerator
+)
 from urllib3.fields import RequestField
 from urllib3.packages.six import b, u
 
 
 BOUNDARY = '!! test boundary !!'
 
 
-class TestIterfields(unittest.TestCase):
+class TestMultipartEncoderGenerator(unittest.TestCase):
+    def test_len(self):
+        fieldsets = [
+            [('k', 'v'), ('k2', 'v2')],
+            [('k', b'v'), (u('k2'), b'v2')],
+            [('k', b'v'), (u('k2'), 'v2')],
+            [('foo', b'a'), ('foo', b'b')],
+        ]
+
+        for fields in fieldsets:
+            encoded, _ = encode_multipart_formdata(fields, boundary=BOUNDARY)
+            encoded = b''.join(encoded)
+            self.assertEqual(len(encoded), len(MultipartEncoderGenerator(fields, boundary=BOUNDARY)))
+
+    def test_memory_usage(self):
+        MEGABYTE = 1024 * 1024
+        data = b'x' * MEGABYTE  # 1 MB of data
+        start_memory = psutil.Process().memory_info().rss
+
+        encoded, _ = encode_multipart_formdata([('k', data)], boundary=BOUNDARY)
+        end_memory = psutil.Process().memory_info().rss
+        self.assertTrue(end_memory <= start_memory + MEGABYTE)
```

Comment:
Yeah, I would tend to agree. In cases where memory usage currently balloons, it's not like we don't know where the usage is, nor that the Python interpreter isn't freeing up unreferenced data; it's simply that we hold onto the reference for two sets of the data at once. Perhaps we could check that the data in the streamer is the same by identity? That'd test that we're not copying the data by mistake.